src/vos/ilog.c

```diff
@@ -960,11 +901,16 @@ ilog_tree_modify(struct ilog_context *lctx, const struct ilog_id *id_in,
 		punchp = (bool *)val_iov.iov_buf;
 		keyp = (struct ilog_id *)key_iov.iov_buf;
 
-		visibility = ilog_status_get(lctx, keyp->id_tx_id,
-					     DAOS_INTENT_UPDATE);
-		if (visibility < 0) {
-			rc = visibility;
-			goto done;
+		visibility = ILOG_UNCOMMITTED;
+
+		if (keyp->id_epoch <= epr->epr_hi &&
+		    keyp->id_epoch >= epr->epr_lo) {
+			visibility = ilog_status_get(lctx, keyp->id_tx_id,
+						     DAOS_INTENT_UPDATE);
+			if (visibility < 0) {
+				rc = visibility;
+				goto done;
+			}
 		}
```

Comment:
Why there is a 'for (;;)' loop in this ilog_tree_modify()?

Suggestion: It seems to me the core logic of "updating in place or adding new entry" are same for embedded tree or real tree, so it could be better for code maintenance if we reorganize the ilog_modify(), ilog_root_migrate() and ilog_tree_modify() a bit (extract this core logic into a common shared function)